### PR TITLE
Fixes for Batchee and Stitchee

### PR DESCRIPTION
**Pull request by indiejames**
## Jira Issue ID
Harmony 1652 2

## Description
This fixes the way we are doing batching to work with Batchee. We had a lot of special case handling for query-cmr that
has been generalized somewhat. Additionally I added some comments explaining how we handle work-item updates
with batching.

## Local Test Steps
Add the following to your services.yml
```
  - name: batchee-stitchee-concise
    description: |
      TODO - fill in description
    data_operation_version: '0.18.0'
    type:
      <<: *default-turbo-config
      params:
        <<: *default-turbo-params
        env:
          <<: *default-turbo-env
          STAGING_PATH: public/asdc-trade/batchee-stitchee-concise
    umm_s: ''
    capabilities:
      concatenation: true
      concatenate_by_default: true
      extend: true
      default_extend_dimensions: ['mirror_step']
      subsetting:
        bbox: false
        variable: false
        temporal: false
      output_formats:
        - application/netcdf4
      reprojection: false
    steps:
      - image: !Env ${QUERY_CMR_IMAGE}
        is_sequential: true
      - image: !Env ${BATCHEE_IMAGE}
        operations: ['concatenate']
      - image: !Env ${STITCHEE_IMAGE}
        operations: ['extend']
      - image: !Env ${PODAAC_CONCISE_IMAGE}
        is_batched: true
        operations: ['concatenate']
        conditional:
          exists: ['concatenate']
```
Add the following to `packages/util/env-defaults`
```
BATCHEE_SERVICE_QUEUE_URLS='["asdc-trade/batchee:latest,http://localstack:4566/queue/batchee.fifo"]'

STITCHEE_SERVICE_QUEUE_URLS='["asdc-trade/stitchee:latest,http://localstack:4566/queue/stitchee.fifo"]'

BATCHEE_IMAGE=asdc-trade/batchee:latest
BATCHEE_REQUESTS_MEMORY=128Mi
BATCHEE_LIMITS_MEMORY=512Mi
BATCHEE_INVOCATION_ARGS='./docker-entrypoint.sh'
BATCHEE_SERVICE_QUEUE_URLS='["asdc-trade/batchee:latest,http://localstack:4566/queue/batchee.fifo"]'

STITCHEE_IMAGE=asdc-trade/stitchee:latest
STITCHEE_REQUESTS_CPU=128m
STITCHEE_LIMITS_CPU=128m
STITCHEE_REQUESTS_MEMORY=128Mi
STITCHEE_LIMITS_MEMORY=10Gi
STITCHEE_INVOCATION_ARGS='./docker-entrypoint.sh'
STITCHEE_SERVICE_QUEUE_URLS='["asdc-trade/stitchee:latest,http://localstack:4566/queue/stitchee.fifo"]'
```

Add the following to `.env`
```
PODAAC_L2_SUBSETTER_LIMITS_MEMORY=2Gi
PODAAC_CONCISE_LIMITS_MEMORY=6Gi
BATCHEE_COLLECTIONS=C1254855648-LARC_CLOUD,C1256510508-LARC_ASDC,C1254854453-LARC_CLOUD
STITCHEE_COLLECTIONS=C1254855648-LARC_CLOUD,C1256510508-LARC_ASDC,C1254854453-LARC_CLOUD
BATCHEE_STITCHEE_CONCISE_COLLECTIONS=C1254855648-LARC_CLOUD,C1256510508-LARC_ASDC,C1254854453-LARC_CLOUD
```

Run the following query:
```
 curl -Ln -bj  "http://localhost:3000/C1254854453-LARC_CLOUD/ogc-api-coverages/1.0.0/collections/all/coverage/rangeset?concatenate=true&skipPreview=true&maxResults=10"
```

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [ ] ~Documentation updated (if needed)~

### CODE REVIEW

File: services/harmony/app/backends/workflow-orchestration/work-item-updates.ts [hunk 1]
```diff
@@ -445,68 +446,80 @@ async function createNextWorkItems(
...
+    } else if (allWorkItemsForStepComplete) {
+      await createAggregatingWorkItem(tx, workItem, nextWorkflowStep, logger);
+      didCreateWorkItem = true;
+    }
+  } else {
+    // Create a new work item for each result using the next step
+    didCreateWorkItem = true;
+
+    // use the sort index from the previous step's work item unless the service was
+    // query-cmr, in which case we start from the previous highest sort index for this step
```
Comment (chris-durbin) [inline services/harmony/app/backends/workflow-orchestration/work-item-updates.ts:489]: Is this comment still correct that it only applies to query-cmr? I guess most importantly does this logic still work for batchee, stitchee, and concise?

**Review by chris-durbin — APPROVED** [review]
I was not able to test this out locally, but paired with James and verified the outputs from Batchee/Stitchee/Concise look correct in Panoply as well as L2SS to Concise.

File: services/harmony/app/backends/workflow-orchestration/work-item-updates.ts [hunk 2]
```diff
@@ -445,68 +446,80 @@ async function createNextWorkItems(
...
+  // 1. next step aggregates AND uses batching - in this case we process each result and put it
+  //  into a batch. As batches fill up we generate a new work-item for the next step and create
+  //  a new batch if neeeded. We have a check to see if we have completed all the work-items for
+  //  the current step, in which case we close the last batch even if it is not full and
+  //  generate a final work-item for the next step
+  // 2. next step aggregates, but does not use batching, so we only create a new work-item
+  //  for the next step if we have completed all the work-items for the current step
+  // 3. next step does not aggregate, so we create a new work-item for the next step for each
+  //  result from this work-item update
+  // if (results && results.length > 0 || nextWorkflowStep.isBatched) {
```
Comment (ygliuvt) [inline services/harmony/app/backends/workflow-orchestration/work-item-updates.ts:459]: remove this line.

**Review by ygliuvt — APPROVED** [review]
Surprised that there is no test. I guess it is due to the difficulty to set up a good test.

**Comment by chris-durbin** [comment]
> Surprised that there is no test. I guess it is due to the difficulty to set up a good test.

Once the service chain is added to services.yml, I think it makes sense to set up a test for it in workflow orchestration. We should add a ticket.

File: services/harmony/app/backends/workflow-orchestration/work-item-updates.ts (hunk 1 above)
Comment (indiejames) [reply services/harmony/app/backends/workflow-orchestration/work-item-updates.ts:489]: query-cmr is the only sequential service we have right now, and we only use that logic for sequential services. To answwer your other question, yes, this works with batchee, stitchee, and concise.

File: services/harmony/app/backends/workflow-orchestration/work-item-updates.ts (hunk 2 above)
Comment (indiejames) [reply services/harmony/app/backends/workflow-orchestration/work-item-updates.ts]: done

**Pull request merged**